janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/vertexcache/CaffeineVertexCache.java

```diff
@@ -0,0 +1,138 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.transaction.vertexcache;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.base.Preconditions;
+import org.jctools.maps.NonBlockingHashMapLong;
+import org.janusgraph.graphdb.internal.InternalVertex;
+import org.janusgraph.graphdb.vertices.AbstractVertex;
+import org.janusgraph.util.datastructures.Retriever;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+public class CaffeineVertexCache implements VertexCache {
+    private static final Logger log =
+        LoggerFactory.getLogger(CaffeineVertexCache.class);
+
+    private final ConcurrentMap<Long, InternalVertex> volatileVertices;
+    private final Cache<Long, InternalVertex> cache;
+
+    private long createdTime; // the timestamp when the cache was created.
+
+    class CaffeineRemovalListener implements RemovalListener<Long, InternalVertex> {
+
+        @Override
+        public void onRemoval(@Nullable Long key,
+                              @Nullable InternalVertex internalVertex,
+                              @Nonnull RemovalCause removalCause) {
+            if (removalCause == RemovalCause.EXPLICIT) {
+                assert volatileVertices.isEmpty();
+                return;
+            }
+            assert (removalCause == RemovalCause.SIZE || removalCause == RemovalCause.REPLACED) : "Cause: " + removalCause;
+            InternalVertex v = internalVertex;
+            if (((AbstractVertex) v).isTxOpen() && (v.isModified() || v.isRemoved())) {
+                volatileVertices.putIfAbsent(key, v);
+            }
+        }
+    }
+
+    public CaffeineVertexCache(final long maxCacheSize, final int initialDirtySize) {
+        volatileVertices = new NonBlockingHashMapLong<>(initialDirtySize);
+        log.debug("Created dirty vertex map with initial size {}", initialDirtySize);
+
+        Caffeine<Long, InternalVertex> cacheBuilder = Caffeine.newBuilder().maximumSize(maxCacheSize)
+            .removalListener(new CaffeineRemovalListener())
+            .recordStats()
```

Comment:
Looks like you forgot to remove `.recordStats()` here. I.e. you are enabling it 2 times, here and below.

Reply:
My fault!

Reply:
My fault!

Reply:
Fixed